lib/internal/errors.js

```diff
@@ -393,6 +378,16 @@ function uvErrmapGet(name) {
   return uvBinding.errmap.get(name);
 }
 
+function captureLargerStackTrace(err) {
```

Comment:
Do we still need this function? It looks like it could be replaced with `Error.captureStackTrace(ex)`.

Reply:
I still need to extend the stack trace size before capturing it. The way the code works is that the stack trace size is set to 0 at the ```Error``` creation (so no stack trace is collected), then it manually invokes ```Error.captureStackTrace(ex)```. That ```captureLargerStackTrace(err)``` is there mainly to centralize at one place extending the stack size and then restoring it back.

Reply:
All invocations of `captureLargerStackTrace()` are made when `Error.stackTraceLimit` is already restored from `0` to the original value, so I still don't get why it's needed. Am I missing something?

Reply:
I need to increase it from the original value to Infinity, since I will be removing some elements and I don't know their number in advance. Then I need to cut off the stack trace to the original value length, so that I can return the exact number of stack trace elements that the user asked for - usually 10.

Reply:
I see now. Then this approach with `Infinity` may lead to increased memory consumption in user applications, especially when they make standard library calls in deeply nested and/or recursive functions.

It can be demonstrated on the following example:
```js
// comment to avoid crashing
Error.stackTraceLimit = Infinity;
function foo(i) {
    if (i === 32000) {
        throw new Error('bar');
    }
    return foo(i + 1);
}
foo(1);
```

If you run it with `node --max-old-space-size=4 --stack-size=32000 recursive-error.js`, it'll crash. Without the `Error.stackTraceLimit` override, the script stops on unhandled error as expected.

Thus, the current implementation seems to be introducing a potential regressions. We could change `Infinity` to `userStackTraceLimit + N` where `N` is the max depth of standard library validation function calls, but that's a bad approach. Do you have an idea for another workaround here?

Reply:
I have another idea, but it is even uglier - rename all functions that are to be hidden to the same name and then use the V8 built-in `Error.captureStackTrace` feature - which is much faster
But it is going to make life even harder for internals developers

Reply:
In fact, I think that aside from some very edge artificial cases - where the heap is smaller than the stack, we should be ok - if you have a call stack that is so large, that it doesn't fit into the heap - you are bound to run into problems anyway

Reply:
Heap may be large enough to fit the call stack, yet it may not have enough free space to allocate an exception as it also contains other JS objects. So, it's better to speak of the free heap space here.

As for the artificial nature of edge cases, the code snippet above is certainly artificial, but there may be modules and user applications that face regressions introduced with this PR. I'm not 100% sure that there will any and I see the certain value introduced by this PR, so it would be nice to get more opinions here.

cc @nodejs/core

Reply:
Proceeding by elimination:
There are only two ways of limiting the captured call stack before it is captured - one is by passing a reference of a function to V8 in `Error.captureStackTrace`. This is not usable without introducing some artificial function that will be able to find on the call stack - this is the old method. In theory, one could imagine some improved solution, where we never add more than once this function - the old method did it multiple times, creating closures at each call.

The other way is limiting the number of entries to a number known in advance - `Error.stackTraceLimit`. Either we estimate the maximum call stack that the Node internals are allowed to reach, either we try to deduce it in a dynamic way.

Reply:
After some thinking, I would personally prefer option 2, i.e. the original solution, but with a modified "padding" (smth like 32 or even 64). Yes, it may lead to trimmed call stacks in produced errors if the padding is not enough, but at least it won't create unpredictable memory pressure.